ironfish/src/network/peers/addressManager.ts

```diff
@@ -2,24 +2,149 @@
+import { ArrayUtils } from '../../utils'
+import { Peer, PeerList } from '..'
+import { ConnectionDirection, ConnectionType } from './connections'
 import { PeerAddress } from './peerAddress'
 
 /**
  * AddressManager stores the necessary data for connecting to new peers
  * and provides functionality for persistence of said data.
  */
 export class AddressManager {
-  hosts: HostsStore
+  hostsStore: HostsStore
 
-  constructor(hostsStore: HostsStore) {
-    this.hosts = hostsStore
+  constructor(files: FileSystem, dataDir?: string) {
+    this.hostsStore = new HostsStore(files, dataDir)
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this.hostsStore.load()
   }
 
   get priorConnectedPeerAddresses(): ReadonlyArray<Readonly<PeerAddress>> {
-    return this.hosts.getArray('priorPeers')
+    return this.hostsStore.getArray('priorPeers')
   }
 
   get possiblePeerAddresses(): ReadonlyArray<Readonly<PeerAddress>> {
-    return this.hosts.getArray('possiblePeers')
+    return this.hostsStore.getArray('possiblePeers')
+  }
+  /**
+   * Adds addresses associated to peers received from peer list
+   */
+  addAddressesFromPeerList(peerList: PeerList): void {
+    const newAddresses: PeerAddress[] = peerList.payload.connectedPeers.map((peer) => ({
+      address: peer.address,
+      port: peer.port,
+      identity: peer.identity ?? null,
+      name: peer.name ?? null,
+    }))
+
+    const possiblePeerStrings = this.possiblePeerAddresses
+      .filter((peerAddress) => !(peerAddress.address == null) && !(peerAddress.port == null))
+      .map(
+        (filteredPeerAddress) =>
+          //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          `${filteredPeerAddress.address!}:${filteredPeerAddress.port!}`,
+      )
+    const possiblePeerSet = new Set(possiblePeerStrings)
+
+    const dedupedAddresses = newAddresses.filter(
+      (newAddress) =>
+        !(newAddress.address == null) &&
+        !(newAddress.port == null) &&
+        !possiblePeerSet.has(`${newAddress.address}:${newAddress.port}`),
+    )
+
+    if (dedupedAddresses.length) {
+      this.hostsStore.set('possiblePeers', [...this.possiblePeerAddresses, ...dedupedAddresses])
+    }
+  }
+
+  /**
+   * Returns a peer address for a disconnected peer by using current peers to
+   * filter out peer addresses. It attempts to find a previously-connected
+   * peer address that is not part of an active connection. If there are none,
+   * then it attempts to find a previously-known peer address.
+   */
+  getRandomDisconnectedPeerAddress(peers: Peer[]): PeerAddress | null {
+    if (
+      this.possiblePeerAddresses.length === 0 &&
+      this.priorConnectedPeerAddresses.length === 0
+    ) {
+      return null
+    }
+
+    const currentPeerAddresses = new Set(
+      //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      peers.filter((peer) => peer.state.identity !== null).map((peer) => peer.state.identity!),
```

Comment:
These are identities rather than addresses, so the name is slightly off here. `flatMap` could be used to avoid the eslint disabling here as well, otherwise I think you could have it take a list of identities rather than peers to avoid having to do the filter here

Reply:
As we're using identities as the sole filter here, I went with the latter option. Thanks!